test/box-luatest/func_test.lua

```diff
@@ -42,3 +43,66 @@ g.test_legacy_opts = function()
         return ret
     end))
 end
+
+g.before_test('test_msgpack_object_args', function()
+    local echo_code = "function(...) return {...} end"
+    local check_code = "function(o) return require('msgpack').is_object(o) end"
+    local decode_code = "function(o) return o:decode() end"
+    g.server:eval("echo = " .. echo_code)
+    g.server:eval("check = " .. check_code)
+    g.server:eval("decode = " .. decode_code)
+    g.server:exec(function(check_code, decode_code)
+        box.schema.func.create('echo')
+        box.schema.func.create('check', {takes_raw_args = true})
+        box.schema.func.create('decode', {takes_raw_args = true})
+        box.schema.func.create(
+            'check_persistent', {body = check_code, takes_raw_args = true})
+        box.schema.func.create(
+            'decode_persistent', {body = decode_code, takes_raw_args = true})
+    end, {check_code, decode_code})
+end)
+
+g.after_test('test_msgpack_object_args', function()
+    g.server:exec(function()
+        box.schema.func.drop('echo')
+        box.schema.func.drop('check')
+        box.schema.func.drop('decode')
+        box.schema.func.drop('check_persistent')
+        box.schema.func.drop('decode_persistent')
+    end)
+    g.server:eval("echo = nil")
+    g.server:eval("check = nil")
+    g.server:eval("decode = nil")
+end)
+
+g.test_msgpack_object_args = function()
+    local args = {'foo', 'bar', {foo = 'bar'}}
+
+    -- remote call
+    local c = net:connect(g.server.net_box_uri)
+    t.assert_equals(c:call('echo', args), args)
+    t.assert(c:call('check', args))
+    t.assert_equals(c:call('decode', args), args)
+    t.assert(c:call('check_persistent', args))
+    t.assert_equals(c:call('decode_persistent', args), args)
+    c:close()
+
+    -- local call
+    local call = function(name, args)
+        return g.server:exec(function(name, args)
+            return box.func[name]:call(args)
+        end, {name, args})
+    end
+    t.assert_equals(call('echo', args), args)
+    t.assert(call('check', args))
+    t.assert_equals(call('decode', args), args)
+    t.assert(call('check_persistent', args))
+    t.assert_equals(call('decode_persistent', args), args)
+
+    -- info
+    t.assert_not(g.server:eval('return box.func.echo.takes_raw_args'))
+    t.assert(g.server:eval('return box.func.check.takes_raw_args'))
+    t.assert(g.server:eval('return box.func.decode.takes_raw_args'))
+    t.assert(g.server:eval('return box.func.check_persistent.takes_raw_args'))
+    t.assert(g.server:eval('return box.func.decode_persistent.takes_raw_args'))
```

Comment:
Lets also add somewhere a test checking that you can return an msgpack object from a function (or is it there already)?
```Lua
function echo(...) return msgpack.object({...}) end
c:call('echo', {1})
```

Reply:
Done.

<details><summary>diff</summary>

```diff
diff --git a/test/box-luatest/func_test.lua b/test/box-luatest/func_test.lua
index 10320f5f174e..69f89b119c40 100644
--- a/test/box-luatest/func_test.lua
+++ b/test/box-luatest/func_test.lua
@@ -45,14 +45,16 @@ g.test_legacy_opts = function()
 end
 
 g.before_test('test_msgpack_object_args', function()
-    local echo_code = "function(...) return {...} end"
+    local echo_code = "function(...) return ... end"
     local check_code = "function(o) return require('msgpack').is_object(o) end"
     local decode_code = "function(o) return o:decode() end"
     g.server:eval("echo = " .. echo_code)
+    g.server:eval("echo_mp = " .. echo_code)
     g.server:eval("check = " .. check_code)
     g.server:eval("decode = " .. decode_code)
     g.server:exec(function(check_code, decode_code)
         box.schema.func.create('echo')
+        box.schema.func.create('echo_mp', {takes_raw_args = true})
         box.schema.func.create('check', {takes_raw_args = true})
         box.schema.func.create('decode', {takes_raw_args = true})
         box.schema.func.create(
@@ -65,12 +67,14 @@ end)
 g.after_test('test_msgpack_object_args', function()
     g.server:exec(function()
         box.schema.func.drop('echo')
+        box.schema.func.drop('echo_mp')
         box.schema.func.drop('check')
         box.schema.func.drop('decode')
         box.schema.func.drop('check_persistent')
         box.schema.func.drop('decode_persistent')
     end)
     g.server:eval("echo = nil")
+    g.server:eval("echo_mp = nil")
     g.server:eval("check = nil")
     g.server:eval("decode = nil")
 end)
@@ -80,7 +84,8 @@ g.test_msgpack_object_args = function()
 
     -- remote call
     local c = net:connect(g.server.net_box_uri)
-    t.assert_equals(c:call('echo', args), args)
+    t.assert_equals({c:call('echo', args)}, args)
+    t.assert_equals(c:call('echo_mp', args), args)
     t.assert(c:call('check', args))
     t.assert_equals(c:call('decode', args), args)
     t.assert(c:call('check_persistent', args))
@@ -93,7 +98,8 @@ g.test_msgpack_object_args = function()
             return box.func[name]:call(args)
         end, {name, args})
     end
-    t.assert_equals(call('echo', args), args)
+    t.assert_equals({call('echo', args)}, args)
+    t.assert_equals(call('echo_mp', args), args)
     t.assert(call('check', args))
     t.assert_equals(call('decode', args), args)
     t.assert(call('check_persistent', args))
@@ -101,6 +107,7 @@ g.test_msgpack_object_args = function()
 
     -- info
     t.assert_not(g.server:eval('return box.func.echo.takes_raw_args'))
+    t.assert(g.server:eval('return box.func.echo_mp.takes_raw_args'))
     t.assert(g.server:eval('return box.func.check.takes_raw_args'))
     t.assert(g.server:eval('return box.func.decode.takes_raw_args'))
     t.assert(g.server:eval('return box.func.check_persistent.takes_raw_args'))
diff --git a/test/box-luatest/net_box_test.lua b/test/box-luatest/net_box_test.lua
index 92e5c3776af1..68b09e287c26 100644
--- a/test/box-luatest/net_box_test.lua
+++ b/test/box-luatest/net_box_test.lua
@@ -76,6 +76,9 @@ end
 
 g.before_test('test_msgpack_object_return', function()
     g.server:eval('function echo(...) return ... end')
+    g.server:eval([[function echo_mp(...)
+        return require('msgpack').object({...})
+    end]])
     g.server:exec(function()
         box.schema.space.create(
             'T', {format = {{'F1', 'unsigned'}, {'F2', 'string'}}})
@@ -85,6 +88,7 @@ end)
 
 g.after_test('test_msgpack_object_return', function()
     g.server:eval('echo = nil')
+    g.server:eval('echo_mp = nil')
     g.server:exec(function()
         box.space.T:drop()
     end)
@@ -101,6 +105,13 @@ g.test_msgpack_object_return = function()
     local ret = c:eval('return ...', {1, 2, 3}, opts)
     t.assert(msgpack.is_object(ret))
     t.assert_equals(ret:decode(), {1, 2, 3})
+    local ret = c:eval("return require('msgpack').object({1, 2, 3})", {}, opts)
+    t.assert(msgpack.is_object(ret))
+    t.assert_equals(ret:decode(), {{1, 2, 3}})
+    -- sic: no return_raw
+    local ret = c:eval("return require('msgpack').object({1, 2, 3})")
+    t.assert_not(msgpack.is_object(ret))
+    t.assert_equals(ret, {1, 2, 3})
 
     -- call
     local ret = c:call('echo', {}, opts)
@@ -109,6 +120,13 @@ g.test_msgpack_object_return = function()
     local ret = c:call('echo', {1, 2, 3}, opts)
     t.assert(msgpack.is_object(ret))
     t.assert_equals(ret:decode(), {1, 2, 3})
+    local ret = c:call('echo_mp', {1, 2, 3}, opts)
+    t.assert(msgpack.is_object(ret))
+    t.assert_equals(ret:decode(), {{1, 2, 3}})
+    -- sic: no return_raw
+    local ret = c:call('echo_mp', {1, 2, 3})
+    t.assert_not(msgpack.is_object(ret))
+    t.assert_equals(ret, {1, 2, 3})
 
     -- dml
     local ret = c.space.T:insert({1, 'foo'}, opts)

```
</details>